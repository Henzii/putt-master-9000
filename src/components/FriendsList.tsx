import { useMutation } from '@apollo/client';
import React, { useEffect, useState } from 'react';
import { Alert, FlatList, Pressable, StyleSheet, View } from "react-native";
import { Avatar, Button, Checkbox, IconButton, Modal, Portal, Text, useTheme } from 'react-native-paper';
import { REMOVE_FRIEND } from '../graphql/mutation';
import useMe from '../hooks/useMe';
import AddFriend from './AddFriend';
import ErrorScreen from './ErrorScreen';
import Loading from './Loading';
import Container from './ThemedComponents/Container';
import { GET_ME_WITH_FRIENDS } from '../graphql/queries';
import SplitContainer from './ThemedComponents/SplitContainer';
import { useNavigate } from 'react-router-native';
import { InitialsAndColors, initialsAndColorGenerator } from '../utils/initialsAndColorGenerator';
import { User } from '../types/user';
import { useBackButton } from './BackButtonProvider';


type FriendListProps = {
    onClick?: (friends: Friend[]) => void,
    hideRemoveButton?: boolean,
    multiSelect?: boolean,
    onBackAction?: () => void
}
export type Friend = {
    id: number | string,
    name: string,
}
const FriendsList = (props: FriendListProps) => {
    const { me, loading, error } = useMe(true);
    const [addFriendModal, setAddFriendModal] = useState(false);
    const [removeFriend] = useMutation(REMOVE_FRIEND, { refetchQueries: [{ query: GET_ME_WITH_FRIENDS }] });
    const [selectedFriends, setSelectedFriends] = useState<Friend[]>([]);
    const navi = useNavigate();
    const backButton = useBackButton();

    useEffect(() => {
        if (props.onBackAction) {
            backButton.setCallBack(props.onBackAction);
            return () => backButton.setCallBack(undefined);
        }
    }, []);

    const handleFriendClick = (friend: Friend) => {
        if (!props.multiSelect) {
            if (props.onClick) {
                props.onClick([friend]);
            }
        }
        else if (!selectedFriends.find(f => f.id === friend.id)) {
            setSelectedFriends(selectedFriends.concat(friend));
        } else {
            setSelectedFriends(selectedFriends.filter(f => f.id !== friend.id));
        }
    };
    const handleOkClick = () => {
        if (props.onClick) props.onClick(selectedFriends);
    };
    const handleKillFriend = (friendId: string | number, friendName?: string) => {
        Alert.alert(
            'Byebye friend',
            `Friends are replaceable but disc golf is for life!\n\nReally remove ${friendName}?`,
            [
                { text: 'Cancel' },
                {
                    text: 'Do it',
                    onPress: () => removeFriend({ variables: { friendId } })
                }
            ]
        );
    };
    if (loading) {
        return <Loading />;
    }
    if (error) {
        return <ErrorScreen errorMessage={error.message} />;
    }

    const initialsAndColors = initialsAndColorGenerator(me?.friends ?? []);

    return (
        <Container noPadding>
            <Portal>
                <Modal
                    visible={addFriendModal}
                    onDismiss={() => setAddFriendModal(false)}
                    contentContainerStyle={{ display: 'flex', justifyContent: 'center', alignItems: 'center' }}
                >
                    <AddFriend onCancel={() => setAddFriendModal(false)} />
                </Modal>
            </Portal>
            <SplitContainer spaceAround>
                <Button icon="plus" onPress={() => setAddFriendModal(true)}>Add friend</Button>
                {!props.multiSelect && <Button icon="plus" onPress={() => navi("/signUp/createFriend")}>Create friend</Button>}
            </SplitContainer>
            <FlatList
                style={tyyli.lista}
                data={me?.friends}
                renderItem={({ item }) => (
                    <SingleFriend
                        onClick={handleFriendClick}
                        onDelete={handleKillFriend}
                        friend={item}
                        selected={selectedFriends.find(f => f.id === item.id) ? true : false}
                        showRemoveButton={!props.hideRemoveButton}
                        showCheckBox={props.multiSelect}
                        initialsAndColors={initialsAndColors}
                    />
                )}
            />
            {props.multiSelect && <Button style={tyyli.button} mode="contained" onPress={handleOkClick}>OK</Button>}
        </Container>
    );
};

type SingleFriendProps = {
    friend: User,
    onClick?: (friend: Friend) => void,
    onDelete?: (id: number | string, name?: string) => void,
    showRemoveButton?: boolean,
    showCheckBox?: boolean,
    selected?: boolean,
    initialsAndColors: InitialsAndColors
}
const SingleFriend = ({ friend, onClick, onDelete, showRemoveButton = true, showCheckBox = false, selected = false, initialsAndColors }: SingleFriendProps) => {
    const { colors } = useTheme();
    const handleFriendClick = () => {
        if (onClick) onClick(friend);
    };
    const handleDelete = () => {
        if (onDelete) onDelete(friend.id as string, friend.name);
    };
    const {label: avatarLabel, color: avatarColor} = initialsAndColors[friend.id];
    return (
        <Pressable onPress={handleFriendClick}>
            <View style={[tyyli.singleFriend, { backgroundColor: colors.surface }, (selected && tyyli.selectedBackground)]}>
                <View style={tyyli.singleFriendName}>
                    <Avatar.Text
                        label={avatarLabel}
                        size={40}
                        labelStyle={{fontSize: avatarLabel.length > 1 ? 18 : 20}}
                        style={{backgroundColor: avatarColor}}
                    />
                    {/*<Avatar.Icon icon={selected ? 'account-tie' : 'account'} size={40} style={{backgroundColor: selected ? colors.surface : '#89ab9f'}} />*/}
                    <Text style={tyyli.singleFriendText}>{friend.name}</Text>
                </View>
                {showRemoveButton && <IconButton iconColor="darkred" icon="trash-can-outline" onPress={handleDelete} style={{backgroundColor: colors.background}} />}
                {showCheckBox && <Checkbox status={selected ? 'checked' : 'unchecked'} />}
            </View>
        </Pressable>
    );
};

const tyyli = StyleSheet.create({
    main: {
        width: '100%',
    },
    selectedBackground: {
        backgroundColor: '#89ab9f',
    },
    singleFriendName: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
    },
    singleFriendText: {
        fontSize: 18,
        marginLeft: 8,
    },
    singleFriend: {
        display: 'flex',
        justifyContent: 'space-between',
        flexDirection: 'row',
        alignItems: 'center',
        borderRadius: 15,
        borderWidth: 1,
        elevation: 5,
        padding: 8,
        marginHorizontal: 10,
        marginVertical: 5,
    },
    button: {
        margin: 10,
    },
    separator: {
        minHeight: 1,
        borderTopWidth: 1,
        borderColor: 'lightgray',
    },
    lista: {
        marginTop: 10,
    },
    otsikko: {
        padding: 10,
        textAlign: 'center',
    }
});
export default FriendsList;