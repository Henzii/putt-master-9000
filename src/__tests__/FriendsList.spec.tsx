import React from 'react';

import { render, waitFor } from '@testing-library/react-native';
import { MockedProvider } from '@apollo/react-testing';
import FriendsList from '../components/FriendsList';
import { Provider } from 'react-native-paper';
import { InMemoryCache } from 'apollo-boost';
import { getMeWithFriendsMock } from './mocks/getMeMock';
import { mockedMe as testiMe } from './mocks/getMeMock';
jest.mock('react-router-native', () => {
    return {
        useNavigate: () => null,
    };
});

const Wrapped = () => (
    <Provider>
        <MockedProvider mocks={[getMeWithFriendsMock]} addTypename={true}
            cache={
                new InMemoryCache({
                  addTypename: false,
                  fragmentMatcher: { match: () => true},
                })
              }
        >
            <FriendsList />
        </MockedProvider>
    </Provider>
);

describe('<FriendList /> testit', () => {
    it('Frendit renderöityy...', async () => {
        const { getByText } = render(<Wrapped />);

        expect(getByText('Loading...')).toBeDefined();
        // Odotetaan että molemmat kaverit löytyy ja että ne myös näkyy
        await waitFor(() => {
            if (!testiMe.friends) return;
            expect(getByText(testiMe.friends[0].name)).toBeDefined();
            expect(getByText(testiMe.friends[1].name)).toBeDefined();
        });
    });

});
