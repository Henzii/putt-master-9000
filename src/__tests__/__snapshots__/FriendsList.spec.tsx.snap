// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FriendList /> testit Frendit renderöityy... 1`] = `
Array [
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          undefined,
          undefined,
          undefined,
          Object {
            "flex": 1,
          },
          Object {
            "backgroundColor": "#f6f6f6",
            "padding": 0,
          },
          undefined,
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 24,
                  "letterSpacing": 0,
                  "lineHeight": 32,
                  "marginVertical": 2,
                },
                Object {
                  "padding": 10,
                  "textAlign": "center",
                },
              ],
            ],
          ]
        }
      >
        Friends
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            },
            undefined,
            undefined,
            Object {
              "justifyContent": "space-around",
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "marginLeft": 12,
                    "marginRight": -4,
                  }
                }
              >
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#6200ee",
                        "fontSize": 16,
                      },
                      Array [
                        Object {
                          "lineHeight": 16,
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "material-community",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  󰐕
                </Text>
              </View>
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#6200ee",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Add friend
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "marginLeft": 12,
                    "marginRight": -4,
                  }
                }
              >
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#6200ee",
                        "fontSize": 16,
                      },
                      Array [
                        Object {
                          "lineHeight": 16,
                          "transform": Array [
                            Object {
                              "scaleX": 1,
                            },
                          ],
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                      ],
                      Object {
                        "fontFamily": "material-community",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  󰐕
                </Text>
              </View>
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#6200ee",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Create friend
              </Text>
            </View>
          </View>
        </View>
      </View>
      <RCTScrollView
        ItemSeparatorComponent={[Function]}
        ListFooterComponent={[Function]}
        data={
          Array [
            Object {
              "blockFriendRequests": false,
              "id": "userid2",
              "name": "TestUser2",
            },
            Object {
              "blockFriendRequests": false,
              "id": "userid3",
              "name": "TestUser3",
            },
          ]
        }
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Object {
            "borderColor": "lightgray",
            "borderTopWidth": 1,
            "marginTop": 10,
          }
        }
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 22,
                    },
                    Object {
                      "backgroundColor": "#ffffff",
                    },
                    false,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "fontFamily": "System",
                            "fontWeight": "400",
                            "writingDirection": "ltr",
                          },
                          Array [
                            Object {
                              "fontSize": 16,
                              "letterSpacing": 0.5,
                              "lineHeight": 24,
                              "marginVertical": 2,
                            },
                            undefined,
                          ],
                        ],
                      ]
                    }
                  >
                    TestUser2
                  </Text>
                </View>
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    Object {
                      "disabled": undefined,
                    }
                  }
                  accessible={true}
                  focusable={true}
                  hitSlop={
                    Object {
                      "bottom": 6,
                      "left": 6,
                      "right": 6,
                      "top": 6,
                    }
                  }
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                          "margin": 6,
                          "overflow": "hidden",
                        },
                        Object {
                          "borderRadius": 18,
                          "height": 36,
                          "width": 36,
                        },
                        undefined,
                        undefined,
                      ],
                    ]
                  }
                >
                  <View>
                    <Text
                      accessibilityElementsHidden={true}
                      allowFontScaling={false}
                      importantForAccessibility="no-hide-descendants"
                      pointerEvents="none"
                      selectable={false}
                      style={
                        Array [
                          Object {
                            "color": "rgb(223,50,50)",
                            "fontSize": 24,
                          },
                          Array [
                            Object {
                              "lineHeight": 24,
                              "transform": Array [
                                Object {
                                  "scaleX": 1,
                                },
                              ],
                            },
                            Object {
                              "backgroundColor": "transparent",
                            },
                          ],
                          Object {
                            "fontFamily": "material-community",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                    >
                      󰩹
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "borderColor": "lightgray",
                  "borderTopWidth": 1,
                  "minHeight": 1,
                }
              }
            />
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "padding": 22,
                    },
                    Object {
                      "backgroundColor": "#ffffff",
                    },
                    false,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "textAlign": "left",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "fontFamily": "System",
                            "fontWeight": "400",
                            "writingDirection": "ltr",
                          },
                          Array [
                            Object {
                              "fontSize": 16,
                              "letterSpacing": 0.5,
                              "lineHeight": 24,
                              "marginVertical": 2,
                            },
                            undefined,
                          ],
                        ],
                      ]
                    }
                  >
                    TestUser3
                  </Text>
                </View>
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    Object {
                      "disabled": undefined,
                    }
                  }
                  accessible={true}
                  focusable={true}
                  hitSlop={
                    Object {
                      "bottom": 6,
                      "left": 6,
                      "right": 6,
                      "top": 6,
                    }
                  }
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "alignItems": "center",
                          "justifyContent": "center",
                          "margin": 6,
                          "overflow": "hidden",
                        },
                        Object {
                          "borderRadius": 18,
                          "height": 36,
                          "width": 36,
                        },
                        undefined,
                        undefined,
                      ],
                    ]
                  }
                >
                  <View>
                    <Text
                      accessibilityElementsHidden={true}
                      allowFontScaling={false}
                      importantForAccessibility="no-hide-descendants"
                      pointerEvents="none"
                      selectable={false}
                      style={
                        Array [
                          Object {
                            "color": "rgb(223,50,50)",
                            "fontSize": 24,
                          },
                          Array [
                            Object {
                              "lineHeight": 24,
                              "transform": Array [
                                Object {
                                  "scaleX": 1,
                                },
                              ],
                            },
                            Object {
                              "backgroundColor": "transparent",
                            },
                          ],
                          Object {
                            "fontFamily": "material-community",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                    >
                      󰩹
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
          >
            <View
              style={
                Object {
                  "borderColor": "lightgray",
                  "borderTopWidth": 1,
                  "minHeight": 1,
                }
              }
            />
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>,
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />,
]
`;
